src/main/java/com/minecolonies/entity/ai/citizen/farmer/EntityAIWorkFarmer.java

```diff
@@ -21,520 +21,550 @@
-        }
-
-        this.incrementActionsDone();
-        return true;
-    }
-
-    /**
-     * Checks if the ground should be hoed and the block above removed.
-     *
-     * @param position the position to check.
-     * @param field    the field close to this position.
-     * @return true if should be hoed.
-     */
-    private boolean shouldHoe(@NotNull final BlockPos position, @NotNull final Field field)
-    {
-        return !field.isNoPartOfField(world, position)
+ {
+     /**
+      * The standard delay the farmer should have.
+      */
+     private static final int     STANDARD_DELAY      = 5;
+     /**
+      * The bonus the farmer gains each update is level/divider.
+      */
+     private static final int     DELAY_DIVIDER       = 10;
+     /**
+      * Changed after finished harvesting in order to dump the inventory.
+      */
+     private              boolean shouldDumpInventory = false;
+     /**
+      * The offset to work at relative to the scarecrow.
+      */
+     @Nullable
+     private BlockPos workingOffset;
+     /**
+      * The delay the farmer should have each action: hoeing, planting, harvesting.
+      */
+     private int workingDelay = STANDARD_DELAY - this.worker.getLevel() / DELAY_DIVIDER;
+
+     /**
+      * Defines if the farmer should request seeds for the current field.
+      */
+     private boolean requestSeeds = true;
+
+     /**
+      * Defines if the farmer should try to get the seeds from his chest.
+      */
+     private boolean shouldTryToGetSeed = true;
+
+     /**
+      * Constructor for the Farmer.
+      * Defines the tasks the Farmer executes.
+      *
+      * @param job a farmer job to use.
+      */
+     public EntityAIWorkFarmer(@NotNull JobFarmer job)
+     {
+         super(job);
+         super.registerTargets(
+                 new AITarget(IDLE, () -> START_WORKING),
+                 new AITarget(START_WORKING, this::startWorkingAtOwnBuilding),
+                 new AITarget(PREPARING, this::prepareForFarming),
+                 new AITarget(FARMER_HOE, this::hoe),
+                 new AITarget(FARMER_PLANT, this::plant),
+                 new AITarget(FARMER_HARVEST, this::harvest)
+         );
+         worker.setSkillModifier(2 * worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
+         worker.setCanPickUpLoot(true);
+     }
+
+     /**
+      * Redirects the fisherman to his building.
+      *
+      * @return the next state.
+      */
+     private AIState startWorkingAtOwnBuilding()
+     {
+         if (walkToBuilding())
+         {
+             return getState();
+         }
+         return PREPARING;
+     }
+
+     /**
+      * Prepares the farmer for farming.
+      * Also requests the tools and checks if the farmer has sufficient fields.
+      *
+      * @return the next AIState
+      */
+     @NotNull
+     private AIState prepareForFarming()
+     {
+         @Nullable BuildingFarmer building = getOwnBuilding();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'building' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)